File: src/FarmAnimal.cpp
#include "../include/FarmAnimal.hpp"
#include <iostream>

using namespace std;

int FarmAnimal::animalCount = 0;

/*
Constructor yang menerima parameter posisi, nama)
*/
FarmAnimal::FarmAnimal(){
    animalCount++;
}

void FarmAnimal::initializeFarmAnimal(pair<int,int> _position, string _name, Land *_landPos){
    position = _position;
    name = _name;
    landPos = _landPos;
}

/*
Destructor untuk mengurangi jumlah hewan.
*/
FarmAnimal::~FarmAnimal(){
    animalCount--;
}

//Getter

//Mendapatkan position
pair<int,int> FarmAnimal::getPosition(){
    return position;
}

//Mendapat waktu sebelum hewan mati
int FarmAnimal::getTimeUntilDead(){
    return timeUntilDead;
}

//Mendapat waktu sebelum hewan lapar
int FarmAnimal::getTimeUntilHungry(){
    return timeUntilHungry;
}

//Mendapat apakah produk sudah diproduksi
bool FarmAnimal::getIsProductProduced(){
    return isProductProduced;
}

//Mendapat nama dari FarmAnimal
string FarmAnimal::getName(){
    return name;
}

//Mendapat Land tempat hewan berada
Land* FarmAnimal::getLandPos(){
    return landPos;
}

//Setter

//Menge-set posisi FarmAnimal
void FarmAnimal::setPosition(pair<int,int> _position){
    position = _position;
}

//Menge-set waktu sebelum hewan mati
void FarmAnimal::setTimeUntilDead(int _timeUntilDead){
    timeUntilDead = _timeUntilDead;
}

//Menge-set waktu sebelum hewan lapar
void FarmAnimal::setTimeUntilHungry(int _timeUntilHungry){
    timeUntilHungry = _timeUntilHungry;
}

//Menge-set apakah produk sudah diproduksi
void FarmAnimal::setIsProductProduced(bool _isProductProduced){
    isProductProduced = _isProductProduced;
}

//Menge-set nama dari FarmAnimal
void FarmAnimal::setName(string _name){
    name = _name;
}

//Menge-set Land tempat hewan berada
void FarmAnimal::setLandPos(Land *_newLand){
    landPos = _newLand;
}

//Method untuk makan rumput di tempat hewan itu berdiri
void FarmAnimal::eat(){
    if ((*landPos).isGrown()){
        timeUntilHungry = TIME_UNTIL_HUNGRY_DEFAULT;
        (*landPos).ungrow();
    }
}

//Bergerak
void FarmAnimal::move(DirectionType direction, LinkedList<LinkedList<Cell*>*>* map){
    switch (direction){
        //Atas
        case UP:
            //Jika masih bisa ke atas
            if (position.first > 0){
                Cell* tempCell = ((*map).get(position.first - 1))->get(position.second);
                if ((*tempCell).isWalkable()){
                    Land* tempLand = dynamic_cast<Land*> (tempCell);
                    if (!(*tempLand).isOccupied()){
                        cout << "BERGERAK KE " << position.first - 1 << " " << position.second << endl;
                        landPos->setObjectHere(NULL);
                        landPos = tempLand;
                        landPos->setObjectHere(this);
                        position.first -= 1;
                    }
                }
            }
            break;
        //Bawah
        case DOWN:
            //Jika masih bisa ke bawah
            if (position.first < (*map).size() - 1){
                Cell* tempCell = ((*map).get(position.first + 1))->get(position.second);
                if ((*tempCell).isWalkable()){
                    Land* tempLand = dynamic_cast<Land*> (tempCell);
                    if (!(*tempLand).isOccupied()){
                        cout << "BERGERAK KE " << position.first + 1 << " " << position.second << endl;
                        landPos->setObjectHere(NULL);
                        landPos = tempLand;
                        landPos->setObjectHere(this);
                        position.first += 1;
                    }
                }
            }
            break;
        //Kiri
        case LEFT:
            //Jika masih bisa ke kiri
            if (position.second > 0){
                Cell* tempCell = ((*map).get(position.first))->get(position.second - 1);
                if ((*tempCell).isWalkable()){
                    Land* tempLand = dynamic_cast<Land*> (tempCell);
                    if (!(*tempLand).isOccupied()){
                        cout << "BERGERAK KE " << position.first << " " << position.second - 1 << endl;
                        landPos->setObjectHere(NULL);
                        landPos = tempLand;
                        landPos->setObjectHere(this);
                        position.second -= 1;
                    }
                }
            }
            break;
        //Kanan
        case RIGHT:
            //Jika masih bisa ke kanan
            if (position.second < (*map).get(position.first)->size() - 1){
                Cell* tempCell = ((*map).get(position.first))->get(position.second + 1);
                if ((*tempCell).isWalkable()){
                    Land* tempLand = dynamic_cast<Land*> (tempCell);
                    if (!(*tempLand).isOccupied()){
                        cout << "BERGERAK KE " << position.first << " " << position.second + 1 << endl;
                        landPos->setObjectHere(NULL);
                        landPos = tempLand;
                        landPos->setObjectHere(this);
                        position.second += 1;
                    }
                }
            }
            break;
    }
}


//Method untuk mengupdate keadaan animal
//Jika isHungry = True, timeUntilDead = timeUntilDead-1
//Jika isHungry = False, timeUntilHungry = timeUntilHungry-1, dan jika timeUntilHungry = 0, isHungry = True
void FarmAnimal::updateCondition(){
    if (isHungry())
        timeUntilDead--;
    else {
        if (timeUntilHungry > 0)
            timeUntilHungry--;
        else
            hungry = true;
    }
}

File: Java/src/engisubs/gameobject/farmanimal/FarmAnimal.java
package engisubs.gameobject.farmanimal;

import java.util.*;
import engisubs.gameobject.*;
import engisubs.gameobject.cell.*;
import engisubs.gameobject.cell.land.*;

abstract public class FarmAnimal extends GameObject {
    protected static int animalCount = 0;
    protected Map <String, Integer> position = new HashMap<>();
    protected boolean hungry = false;
    protected int timeUntilDead = 10;
    protected int timeUntilHungry = 0;
    protected boolean isProductProduced = true;
    protected String name = "";
    protected Land landPos = null;
    protected boolean killable = false;

    protected void initializeFarmAnimal(Map <String, Integer> _position, String _name, Land _landPos){
        position.put("Row", _position.get("Row"));
        position.put("Col", _position.get("Col"));
        name = _name;
        landPos = _landPos;
    }

    public Map <String, Integer> getPosition(){
        return position;
    }

    public int getTimeUntilDead(){
        return timeUntilDead;
    }

    public int getTimeUntilHungry(){
        return timeUntilHungry;
    }

    public boolean getIsProductProduced(){
        return isProductProduced;
    }

    public String getName(){
        return name;
    }

    public Land getLandPos(){
        return landPos;
    }

    public boolean isKillable(){
        return killable;
    }

    public boolean isHungry(){
        return hungry;
    }

    public abstract void talk();

    protected void setPosition(Map <String, Integer> _newPosition){
        position.replace("Row", _newPosition.get("Row"));
        position.replace("Col", _newPosition.get("Col"));
    }

    protected void setTimeUntilDead(int _timeUntilDead){
        timeUntilDead = _timeUntilDead;
    }

    protected void setIsProductProduced(boolean _isProductProduced){
        isProductProduced = _isProductProduced;
    }

    public void eat(){
        if (landPos.isGrown()) {
            System.out.println("MAKAN VALID");
            timeUntilDead = 10;
            timeUntilHungry = 10;
            landPos.ungrow();
            setIsProductProduced(false);
        }
    }

    /*
        Method untuk mengupdate keadaan animal
        Jika isHungry = True, timeUntilDead = timeUntilDead-1
        Jika isHungry = False, timeUntilHungry = timeUntilHungry-1, dan jika timeUntilHungry = 0, isHungry = True
    */
    public void updateCondition(){
        if (isHungry()){
            timeUntilDead -= 1;
        }else {
            if (timeUntilHungry > 0)
                timeUntilHungry--;
            else
                hungry = true;
        }
    }

    public void move(DirectionType dir, List<LinkedList<Cell>> map){
        switch (dir){
            //Atas
            case UP:
                //Jika masih bisa ke atas
                if (position.get("Row") > 0){
                    Cell tempCell = (map.get(position.get("Row") - 1)).get(position.get("Col"));
                    if (tempCell.isWalkable()){
                        Land tempLand = (Land) tempCell;
                        if (!tempLand.isOccupied()){
                            //cout << "BERGERAK KE " << position.first - 1 << " " << position.second << endl;
                            landPos.setObjectHere(null);
                            landPos = tempLand;
                            landPos.setObjectHere(this);
                            position.replace("Row", position.get("Row") - 1);
                        }
                    }
                }
                break;
            //Bawah
            case DOWN:
                //Jika masih bisa ke bawah
                if (position.get("Row") < map.size() - 1){
                    Cell tempCell = (map.get(position.get("Row") + 1)).get(position.get("Col"));
                    if (tempCell.isWalkable()){
                        Land tempLand = (Land) tempCell;
                        if (tempLand.isOccupied()){
                            //cout << "BERGERAK KE " << position.first + 1 << " " << position.second << endl;
                            landPos.setObjectHere(null);
                            landPos = tempLand;
                            landPos.setObjectHere(this);
                            position.replace("Row", position.get("Row") - 1);
                        }
                    }
                }
                break;
            //Kiri
            case LEFT:
                //Jika masih bisa ke kiri
                if (position.get("Col") > 0){
                    Cell tempCell = (map.get(position.get("Row"))).get(position.get("Col") - 1);
                    if (tempCell.isWalkable()){
                        Land tempLand = (Land) tempCell;
                        if (!tempLand.isOccupied()){
                            //cout << "BERGERAK KE " << position.first << " " << position.second - 1 << endl;
                            landPos.setObjectHere(null);
                            landPos = tempLand;
                            landPos.setObjectHere(this);
                            position.replace("Col", position.get("Col") - 1);
                        }
                    }
                }
                break;
            //Kanan
            case RIGHT:
                //Jika masih bisa ke kanan
                if (position.get("Col") < map.get(position.get("Row")).size() - 1){
                    Cell tempCell = (map.get(position.get("Row"))).get(position.get("Col") + 1);
                    if (tempCell.isWalkable()){
                        Land tempLand = (Land) tempCell;
                        if (!tempLand.isOccupied()){
                            //cout << "BERGERAK KE " << position.first << " " << position.second + 1 << endl;
                            landPos.setObjectHere(null);
                            landPos = tempLand;
                            landPos.setObjectHere(this);
                            position.replace("Col", position.get("Col") + 1);
                        }
                    }
                }
                break;
        }
    }
}